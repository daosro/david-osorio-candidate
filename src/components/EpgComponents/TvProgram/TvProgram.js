import React, { useCallback, useEffect, useMemo, useState } from "react";
import clsx from "clsx";
import { useNavigate  } from "react-router-dom";
import { generatePath } from "react-router";

import {
  DATE_PATTERN_HHMM,
  DATE_PATTERN_YYYY_MM_DD_T,
} from "../../../constants/date";
import {
  formatDateStringToPattern,
  getMinutesBetweenDates,
  isDateBeforeNow,
  isNowBetweenDates,
} from "../../../utils/date";

import useStyles from "./TvProgram.style";
import { LIVE_PATH, VODP_PATH } from "../../../constants/routes";

const TvProgram = ({
  title,
  id,
  start,
  end,
  duration = getMinutesBetweenDates(start, end, DATE_PATTERN_YYYY_MM_DD_T),
}) => {
  const navigate = useNavigate ();
  const [isLiveProgram, setIsLiveProgram] = useState(false);
  const [isProgramOver, setIsProgramOver] = useState(false);
  const classes = useStyles({ duration, useLink: isLiveProgram || isProgramOver });

  /**
   * Check if program is live every minute to update the styles,
   * I know that is not the best way to do it, with more time I would calculate the remaining time to start or end
   * the program and update the styles accordingly to the remaining time not every minute.
   */
  useEffect(() => {
    let interval;
    const isProgramEnded = isDateBeforeNow(end, DATE_PATTERN_YYYY_MM_DD_T);
    if (!isProgramEnded) {
      interval = setInterval(() => {
        setIsLiveProgram(
          isNowBetweenDates(start, end, DATE_PATTERN_YYYY_MM_DD_T)
        );
        const isProgramEnded = isDateBeforeNow(end, DATE_PATTERN_YYYY_MM_DD_T);
        setIsProgramOver(isProgramEnded);
        if (isProgramEnded) {
          clearInterval(interval);
        }
      }, 1000);
    }
    setIsProgramOver(isProgramEnded);
    return () => clearInterval(interval);
  }, [start, end]);

  const { startTime, endTime } = useMemo(
    () => ({
      startTime: formatDateStringToPattern(
        start,
        DATE_PATTERN_YYYY_MM_DD_T,
        DATE_PATTERN_HHMM
      ),
      endTime: formatDateStringToPattern(
        end,
        DATE_PATTERN_YYYY_MM_DD_T,
        DATE_PATTERN_HHMM
      ),
    }),
    [start, end]
  );

  const handleClick = useCallback(() => {
    if (isLiveProgram || isProgramOver) {
      navigate(
        isProgramOver
          ? generatePath(VODP_PATH, { programId: id })
          : generatePath(LIVE_PATH, { channelId: id })
      );
    }
  }, [id, isLiveProgram, isProgramOver]);

  return (
    <div
      className={clsx(classes.root, { [classes.live]: isLiveProgram })}
      onClick={handleClick}
    >
      <div className={classes.content}>
        <div className={classes.title}>{title}</div>
        <div className={classes.date}>
          {startTime} - {endTime}
        </div>
      </div>
    </div>
  );
};

export default TvProgram;
