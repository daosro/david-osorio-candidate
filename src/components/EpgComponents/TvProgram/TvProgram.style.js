import { createUseStyles } from "react-jss";

export default createUseStyles(
  {
    root: {
      display: "flex",
      borderRight: "1px solid",
      flexDirection: "column",
      backgroundColor: "#000000",
      borderRightColor: "#393939",
      width: props => `calc(((400px / 60) * ${props.duration}) - 1px)`,
      cursor: props => props.useLink ? "pointer" : "default",
    },
    live: {
      backgroundColor: "#393939",
      transition: 'transform .5s',
      '&:hover': {
        transform: 'scale(1.1)',
      }
    },
    content: {
      padding: "0.5rem",
    },
    title: {
      fontSize: "13px",
    },
    date: {
      color: "#939393",
      fontSize: "12px",
    },
  },
  { name: "TvProgram" }
);
