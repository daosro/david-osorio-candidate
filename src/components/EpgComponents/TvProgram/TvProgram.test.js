import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import TvProgram from "./TvProgram";

describe("TvProgram", () => {
  it("should render correctly", () => {
    const { container } = render(
      <MemoryRouter>
        <TvProgram />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
  it("should contain the title", () => {
    render(
      <MemoryRouter>
        <TvProgram title="dummy_program" />
      </MemoryRouter>
    );
    expect(screen.getByText("dummy_program")).toBeInTheDocument();
  });
});