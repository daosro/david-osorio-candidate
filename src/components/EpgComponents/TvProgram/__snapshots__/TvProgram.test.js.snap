// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TvProgram should render correctly 1`] = `
<div>
  <div
    class="TvProgram-root-0-2-1 TvProgram-root-d0-0-2-6"
  >
    <div
      class="TvProgram-content-0-2-3"
    >
      <div
        class="TvProgram-title-0-2-4"
      />
      <div
        class="TvProgram-date-0-2-5"
      >
        Invalid date
         - 
        Invalid date
      </div>
    </div>
  </div>
</div>
`;
